### PR TITLE
WINDUP-974: Support queries based upon annotation values

### PR DESCRIPTION
**Pull request by jsight**
https://issues.jboss.org/browse/WINDUP-974

### CODE REVIEW

**Comment by OndraZizka** [comment]
Nice. Seems OK to me.

File: rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/condition/annotation/AnnotationListCondition.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import org.jboss.windup.rules.apps.java.scan.ast.annotations.JavaAnnotationTypeValueModel;
+import org.ocpsoft.rewrite.context.EvaluationContext;
+
+import com.google.common.collect.Iterables;
+
+/**
+ * @author <a href="mailto:jesse.sightler@gmail.com">Jesse Sightler</a>
+ */
+public class AnnotationListCondition extends AnnotationCondition
+{
```
Comment (OndraZizka) [inline rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/condition/annotation/AnnotationListCondition.java]: Some javadoc, please.

File: rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/condition/annotation/AnnotationTypeCondition.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import org.jboss.windup.rules.apps.java.scan.ast.annotations.JavaAnnotationTypeValueModel;
+import org.ocpsoft.rewrite.context.EvaluationContext;
+import org.ocpsoft.rewrite.param.ParameterizedPatternResult;
+import org.ocpsoft.rewrite.param.RegexParameterizedPatternParser;
+
+/**
+ * @author <a href="mailto:jesse.sightler@gmail.com">Jesse Sightler</a>
+ */
+public class AnnotationTypeCondition extends AnnotationCondition
+{
```
Comment (OndraZizka) [inline rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/condition/annotation/AnnotationTypeCondition.java]: Some javadoc, please, perhaps with an example of the annotation code to match.

**Comment by jsight** [comment]
Good finds... I'll plan to tweak the javadoc tomorrow.

File: rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/condition/annotation/AnnotationListCondition.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
+package org.jboss.windup.rules.apps.java.condition.annotation;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.jboss.windup.config.GraphRewrite;
+import org.jboss.windup.config.condition.EvaluationStrategy;
+import org.jboss.windup.rules.apps.java.scan.ast.annotations.JavaAnnotationListTypeValueModel;
+import org.jboss.windup.rules.apps.java.scan.ast.annotations.JavaAnnotationLiteralTypeValueModel;
```
Comment (mareknovotny) [inline rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/condition/annotation/AnnotationListCondition.java]: This seems unused

File: rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/condition/annotation/AnnotationTypeCondition.java (hunk 2 above)
Comment (mareknovotny) [inline rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/condition/annotation/AnnotationTypeCondition.java]: we probably need to add docs too in wiki. As this needs to be clear for end users how to use it properly.

File: rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/condition/annotation/AnnotationListCondition.java (hunk 3 above)
Comment (jsight) [inline rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/condition/annotation/AnnotationListCondition.java]: Good catch.

File: rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/condition/annotation/AnnotationTypeCondition.java (hunk 2 above)
Comment (jsight) [inline rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/condition/annotation/AnnotationTypeCondition.java]: Yes, I'll file a jira on the docs with the initial notes once the code is tested and merged.

**Comment by jsight** [comment]
I believe that I have added the missing javadoc.

**Comment by mareknovotny** [comment]
Do I understand correctly that `@Field(store = Store.YES, index = Index.YES, indexNullAs = "-1")`
should be caught by 

```xml
<javaclass references="org.hibernate.search.annotations.Field">
    <location>ANNOTATION</location>
    <annotation-type pattern="indexNullAs" />
<javaclass>
```

**Comment by jsight** [comment]
@mareknovotny - In your example, indexNullAs is an annotation element (java terminology).

I think you really want:

```xml
<javaclass references="org.hibernate.search.annotations.Field">
    <location>ANNOTATION</location>
    <annotation-literal pattern="name" pattern="{*}" />
<javaclass>
```

**Comment by mareknovotny** [comment]
that is what confuses me that xml elements are not descriptive or I can't imagine what they are mapped to java terminology.
I tried this as your example had twice pattern attribute, but without luck.
```xml
<javaclass references="org.hibernate.search.annotations.Field">
    <location>ANNOTATION</location>
    <annotation-literal pattern="indexNullAs"/>
<javaclass>
```

**Comment by jsight** [comment]
Sorry, that should have been name not pattern. element names are not element values. :)

**Comment by jsight** [comment]
I have updated the example.

File: rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/condition/annotation/AnnotationListCondition.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+ * <p>
+ * Matches on an item in an array within an annotation.
+ * </p>
+ *
+ * <p>
+ * For example, given the following annotation:
+ * </p>
+ *
+ * <pre>
+ * @Foo(value = [ 1, 2, 3, 4 ])
```
Comment (Ladicek) [inline rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/condition/annotation/AnnotationListCondition.java]: Array literals use curly braces :-)

Comment (jsight) [inline rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/condition/annotation/AnnotationListCondition.java]: Of course... fixed. :)

**Pull request merged**